### PR TITLE
Decouple script reduction from signing

### PR DESCRIPTION
**Pull request by aslesarenko**
- [x] ReducedTransaction introduced along with methods to create and sign it
- [x] Implemented ColdBlockchainContext

### CODE REVIEW

**Review by MrStahlfelge — APPROVED** [review]
This works perfect, [tested on Android 7 and 11](https://github.com/MrStahlfelge/ergo-wallet-android/commit/c48090a8914f8f668f461848760f72633156c832) with token sending.

With a single inbox and a token, the byte array is around 400 bytes and this is perfect for QR code. 
The only problem I've found is that SignedTransaction.toJson(false, false) does not work and outputs an error - only beautified json can be outputted for the time being. For the said transaction, the json is around 2300 bytes (1900 when minified will work)

**Comment by aslesarenko** [comment]
> SignedTransaction.toJson(false, false)

Why it is necessary?

**Comment by MrStahlfelge** [comment]
> > SignedTransaction.toJson(false, false)
> 
> Why it is necessary?

I used it to serialize the signed transaction to transfer it back to the hot device ([code](https://github.com/MrStahlfelge/ergo-wallet-android/blob/c48090a8914f8f668f461848760f72633156c832/app/src/main/java/org/ergoplatform/android/ErgoFacade.kt#L190)). Because this is done via QR code, too, the size does matter. It is not mandatory to be minified because I could zip the json string as well, but I just stumbled upon that only pretty printed json output works and wanted to note it.

File: appkit/src/test/scala/org/ergoplatform/appkit/TxBuilderSpec.scala [hunk 1]
```diff
@@ -227,4 +228,65 @@ class TxBuilderSpec extends PropSpec with Matchers
...
+        senders, pkContract, amountToSend,
+        new util.ArrayList[ErgoToken]())
+
+      val prover = ctx.newProverBuilder.build // prover without secrets
+      val reduced = prover.reduce(unsigned, 0)
+      reduced should not be(null)
+      reduced
+    }
+
+    // TODO implement serialization/deserialization of ReducedTransaction
```
[the remark this answers is not in the document]
Comment (aslesarenko) [reply appkit/src/test/scala/org/ergoplatform/appkit/TxBuilderSpec.scala:256]: done

File: appkit/src/test/scala/org/ergoplatform/appkit/TxBuilderSpec.scala [hunk 2]
```diff
@@ -227,4 +228,65 @@ class TxBuilderSpec extends PropSpec with Matchers
...
+    val reducedTx = reduced.asInstanceOf[ReducedTransactionImpl].getTx
```
[the remark this answers is not in the document]
Comment (aslesarenko) [reply appkit/src/test/scala/org/ergoplatform/appkit/TxBuilderSpec.scala:257]: done

File: appkit/src/test/scala/org/ergoplatform/appkit/TxBuilderSpec.scala [hunk 3]
```diff
@@ -227,4 +228,65 @@ class TxBuilderSpec extends PropSpec with Matchers
...
+        ctx.getBoxesById("d47f958b201dc7162f641f7eb055e9fa7a9cb65cc24d4447a10f86675fc58328"),
+        exceptionLike[NotImplementedError]())
+
+      // create prover with secrets in the cold context
+      val prover = BoxOperations.createProver(ctx,
+        new File("storage/E2.json").getPath, "abc")
+        .build
+
+      // sign with the cold prover
+      val deserializedTx = ReducedErgoLikeTransactionSerializer.fromBytes(reducedTxBytes)
```
[the remark this answers is not in the document]
Comment (aslesarenko) [reply appkit/src/test/scala/org/ergoplatform/appkit/TxBuilderSpec.scala:280]: implemented ctx.parseReducedTransaction

**Comment by aslesarenko** [comment]
> SignedTransaction.toJson(false, false)

Implemented serialization for SignedTransaction in BlockchainContext.parseSignedTransaction

**Review by MrStahlfelge — COMMENTED** [review]
Implemented and tested the changes on branch https://github.com/MrStahlfelge/ergo-wallet-android/commits/cold-wallet-oc - works perfect.

The only thing still left for me is to obtain the outboxes information from the `ReducedTransaction` to show the users what they are going to sign.

**Comment by MrStahlfelge** [comment]
Okay, found out: `reducedTx.tx.unsignedTx().outputCandidates()` gives a list with the needed information

**Pull request merged**